buildkite/bazelci.py

```diff
@@ -2130,6 +2148,112 @@ def expand_test_target_patterns(bazel_binary, platform, test_targets):
+):
+    print_collapsed_group(
+        f":scissors: Filtering targets that haven't been affected since {diffbase}"
+    )
+
+    tmpdir = tempfile.mkdtemp()
+    eprint(f"Downloading bazel-diff to {tmpdir}")
+    bazel_diff_path = download_bazel_diff(tmpdir)
+    resolved_diffbase = resolve_diffbase(diffbase)
+
+    eprint(f"Running bazel-diff for {resolved_diffbase} and {git_commit}")
+    try:
+        affected_targets = run_bazel_diff(
+            bazel_diff_path, workspace_dir, bazel_binary, resolved_diffbase, git_commit, tmpdir
+        )
+    finally:
+        shutil.rmtree(tmpdir)
+
+    filtered_targets = list(set(expanded_test_targets).intersection(affected_targets))
+    if len(filtered_targets) < len(expanded_test_targets):
+        execute_command(
+            [
+                "buildkite-agent",
+                "annotate",
+                "--style=info",
+                "'This run only contains test targets that have been changed since "
+                "{} due to the {} env variable'".format(resolved_diffbase, USE_BAZEL_DIFF_ENV_VAR),
+            ]
+        )
+
+    return filtered_targets
+
+
+def resolve_diffbase(diffbase):
+    if diffbase in AUTO_DIFFBASE_VALUES:
+        return "HEAD^"
+    elif COMMIT_RE.fullmatch(diffbase):
+        return diffbase
+
+    raise BuildkiteException(
+        "Invalid value '{}' for {} env variable. Must be a Git commit hash or one of {}".format(
+            diffbase, ", ".join(AUTO_DIFFBASE_VALUES)
+        )
+    )
+
+
+def download_bazel_diff(directory):
+    # TODO fweikert: Include bazel-diff in the Docker images if we decide to productionize the experiment.
+    url = "https://github.com/Tinder/bazel-diff/releases/download/4.5.0/bazel-diff_deploy.jar"
+    local_path = os.path.join(directory, "bazel-diff.jar")
+    try:
+        execute_command(["curl", "-sSL", url, "-o", local_path])
+    except subprocess.CalledProcessError as ex:
+        raise BuildkiteInfraException("Failed to download {}, error message:\n%s".format(url, ex))
+    return local_path
+
+
+def run_bazel_diff(
+    bazel_diff_path, workspace_dir, bazel_binary, start_commit, end_commit, data_dir
+):
+    before_json = os.path.join(data_dir, "before.json")
+    after_json = os.path.join(data_dir, "after.json")
+    targets_file = os.path.join(data_dir, "targets.txt")
+
+    try:
+        for commit, json_path in ((start_commit, before_json), (end_commit, after_json)):
+            execute_command(["git", "-C", workspace_dir, "checkout", commit, "--quiet"])
+            execute_command(
+                [
+                    "java",
+                    "-jar",
+                    bazel_diff_path,
+                    "generate-hashes",
+                    "-w",
+                    workspace_dir,
+                    "-b",
+                    bazel_binary,
+                    json_path,
+                ]
+            )
+
+        execute_command(
+            [
+                "java",
```

Comment:
Do you know how bazel-diff works underlying? I don't see we are specifying a target scope it should check. Will it cause Bazel to fetch more external dependencies than needed?

Reply:
Good question. The help text under https://github.com/Tinder/bazel-diff#generate-hashes-command mentions external repos.
I didn't see any fetches when running the code locally, and it was very quick.

Reply:
SG

Reply:
I have another concern, what if we just change the presubmit.yml file to add a test flag, or the .bazelrc file, how does bazel-diff decide which tests to rerun?

Reply:
Maybe we should have a list of files that should cause the pipeline to skip  bazel-diff if touched?

Reply:
Yeah, and maybe we should also pass the test flags to bazel-diff.

Reply:
I'd move that to a future PR, though